File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    //Two Binary Search
    let ROWS = matrix.length
    let COLS = matrix[0].length
    
    //1. Get the target Array/Row
    let topRow = 0
    let botRow = ROWS - 1
    
    while(topRow <= botRow){
        let currRow = Math.floor((topRow + botRow) / 2)
        
        if(target > matrix[currRow][COLS - 1])
            {topRow = currRow + 1}
        else if(target < matrix[currRow][0]) 
            {botRow = currRow - 1}
        else {break}
    }
    
    if (topRow > botRow){
        return false
    }
    
    //2. Get the target Number
    let low = 0
    let high = COLS - 1
    let currRow = Math.floor((topRow + botRow) / 2)
    
    while(low <= high){
        let mid = Math.floor((high + low) / 2)
        
        if(target > matrix[currRow][mid]){low = mid + 1}
        else if(target < matrix[currRow][mid]){high = mid - 1}
        else {return true}
    }
    return false
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix or target is None:
            return False
        # 2D Binary Search
        rows, cols = len(matrix), len(matrix[0])
        lRow, rRow = 0, rows-1
        
        while lRow <= rRow:
            midRow = lRow + ((rRow - lRow)//2)
            
            if matrix[midRow][0] > target:
                rRow = midRow - 1
            elif matrix[midRow][-1] < target:
                lRow = midRow + 1
            else:
                # Element row is Found
                break
        
        if not(lRow <= rRow):
            return False
        
        l,r = 0, cols-1
        row = lRow + ((rRow - lRow)//2)
        
        while l <=r:
            mid = l + ((r - l)//2)
            
            if matrix[row][mid] > target:
                r = mid - 1
            elif matrix[row][mid] < target:
                l = mid + 1
            else:
                return True
        return False
        